File: lib/exports/assessments/questions/slider.rb
# frozen_string_literal: true

module Exports
  module Assessments
    module Questions
      class Slider < Base
        # FROM:
        #   [{
        #     "index": 0,
        #     "value": 12
        #   }, ...]
        # TO:
        #   [12, ...]
        def self.result(user_result, question, scoring = false, _export_with_labels = false) # rubocop:disable Metrics/CyclomaticComplexity,Metrics/PerceivedComplexity
          answers = get_answers(user_result, question) || []
          factors_scoring = question.detect_specified_scoring.
                            each_with_object({}) { |s, sum| sum[s['index']] = s['value']; }
          not_applicable = get_not_applicable(user_result, question) || {}
          na_label = question.props['notApplicableLabel'] || NOT_APPLICABLE_PLACEHOLDER

          answers = Array.new(question.props['choices'].to_i) do |c|
            a = answers.find do |answer|
              answer['index'] == c
            end || (not_applicable[c.to_s] ? { 'value' => na_label } : { 'value' => nil })
            if a['value'].is_a?(Numeric)
              ((scoring && factors_scoring[a['index']]) || 1) * a['value']
            else
              (a['value'] || '')
            end
          end

          answers = Array.new(question_headers_except_duration_size(question)) { '' } if answers.empty?
          answers << get_duration(user_result, question)
          Utility::Array.ensure_size(answers, question_header_size(question))
        end

        def self.question_id_and_choice_headers(question)
          question_id_header = []
          question_choices_header = []
          question.props['choices'].to_i.times do |c|
            question_id_header << "QID#{question.id}_#{c + 1}"
            question_choices_header << question.props.dig('choicesTexts', c)
          end
          append_duration_header(question, question_id_header, question_choices_header)

          { question_id_header: question_id_header, question_choice_header: question_choices_header }
        end
      end
    end
  end
end

File: app/frontend/libs/survey/models/ScoringModules/Slider.js
import _ from 'lodash'
import BaseScoringModule from './BaseScoringModule'

class Slider extends BaseScoringModule {
  fill (question) {
    _.times(question.props.choices, (index) => {
      this.scoring.props.push({ index, value: index + 1 })
    })
  }

  changeValue (index, value) {
    if (value && value >= 0) {
      const object = _.find(this.scoring.props, { index })
      if (object) {
        object.value = value
      } else {
        this.scoring.props.push({ index, value })
      }
    } else {
      _.remove(this.scoring.props, { index })
    }
  }

  toggleReverse (index) {
    const object = _.find(this.scoring.props, { index })
    if (!object) { return }
    if (object.reverse) {
      _.unset(object, 'reverse')
    } else {
      _.set(object, 'reverse', true)
    }
  }

  toggleValue (index) {
    const object = _.find(this.scoring.props, { index })
    if (object) {
      _.remove(this.scoring.props, { index })
    } else {
      this.scoring.props.push({ index, value: '1' })
    }
  }

  toggle (index, reverse) {
    if (reverse) {
      this.toggleReverse(index)
    } else {
      this.toggleValue(index)
    }
  }
}

export default Slider
